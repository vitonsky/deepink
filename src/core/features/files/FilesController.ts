import { z } from 'zod';
import { PGLiteDatabase } from '@core/storage/database/pglite/PGLiteDatabase';
import { qb } from '@utils/db/query-builder';
import { wrapDB } from '@utils/db/wrapDB';

import { IFilesStorage } from '.';

// TODO: add tests
// TODO: add runtime validation
// TODO: implement interface and use interface instead of class
/**
 * Files manager for local database
 */
export class FilesController {
	private db;
	private fileController;
	private readonly workspace;
	constructor(db: PGLiteDatabase, fileController: IFilesStorage, workspace: string) {
		this.db = db;
		this.fileController = fileController;
		this.workspace = workspace;
	}

	public async add(file: File) {
		const db = wrapDB(this.db.get());

		// Insert in DB
		const {
			rows: [{ id: fileId }],
		} = await db.query(
			qb.sql`INSERT INTO files (workspace_id,name,mimetype) VALUES (${this.workspace},${file.name},${file.type}) RETURNING id`,
			z.object({ id: z.string() }),
		);

		// TODO: delete entry in DB if can't upload file. Or try to upload first and then add file to DB
		// TODO: encrypt file
		// Write file
		const buffer = await file.arrayBuffer();
		await this.fileController.write(this.getFilePath(fileId), buffer);

		return fileId;
	}

	public async get(id: string) {
		const db = wrapDB(this.db.get());

		// Insert in DB
		const [fileEntry] = await db
			.query(
				qb.sql`SELECT * FROM files WHERE workspace_id=${this.workspace} AND id=${id}`,

				z.object({
					name: z.string(),
					mimetype: z.string(),
				}),
			)
			.then((result) => result.rows);

		if (!fileEntry) return null;

		const { name, mimetype } = fileEntry;

		const buffer = await this.fileController.get(this.getFilePath(id));
		if (!buffer) return null;

		// TODO: decrypt file
		return new File([buffer], name, { type: mimetype });
	}

	// TODO: remove attached files
	public async delete(filesId: string[]) {
		const db = wrapDB(this.db.get());

		if (filesId.length === 0) return;

		// Delete in database
		await db.query(
			qb.sql`DELETE FROM files WHERE workspace_id=${
				this.workspace
			} AND id IN (${qb.values(filesId)})`,
		);

		// Delete files
		await this.fileController.delete(
			filesId.map((filename) => this.getFilePath(filename)),
		);
	}

	public async query() {
		const db = wrapDB(this.db.get());

		return await db
			.query(
				qb.sql`SELECT id, name, mimetype FROM files WHERE workspace_id=${this.workspace}`,
				z.object({
					id: z.string(),
					name: z.string(),
					mimetype: z.string(),
				}),
			)
			.then(({ rows }) => rows);
	}

	private getFilePath(filename: string) {
		return [this.workspace, filename].join('/');
	}
}
