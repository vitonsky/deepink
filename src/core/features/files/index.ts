// TODO: add method `has`
/**
 * Interface for persistent files storage
 * File name is UUID, not a file system path
 *
 * For desktop app it may be FS or cloud service, for web it may be API or indexedDB
 */
export type IFilesStorage = {
	write: (uuid: string, buffer: ArrayBuffer) => Promise<void>;
	get: (uuid: string) => Promise<ArrayBuffer | null>;
	delete: (uuid: string[]) => Promise<void>;
	list: () => Promise<string[]>;
};

export type IFileController = {
	write: (buffer: ArrayBuffer) => Promise<void>;
	get: () => Promise<ArrayBuffer | null>;
	delete: () => Promise<void>;
};
