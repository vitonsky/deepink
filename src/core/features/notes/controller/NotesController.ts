/* eslint-disable camelcase */
import { Query } from 'nano-queries/core/Query';
import { z } from 'zod';
import { PGLiteDatabase } from '@core/storage/database/pglite/PGLiteDatabase';
import { qb } from '@utils/db/query-builder';
import { wrapDB } from '@utils/db/wrapDB';

import { INote, INoteContent, NoteId } from '..';
import {
	INotesController,
	NoteMeta,
	NotesControllerFetchOptions,
	NoteSortField,
} from '.';

const RowScheme = z
	.object({
		id: z.string(),
		title: z.string(),
		text: z.string(),
		created_at: z.number(),
		updated_at: z.number(),
		history_disabled: z.number(),
		visible: z.number(),
	})
	.transform(
		({
			id,
			title,
			text,
			created_at,
			updated_at,
			history_disabled,
			visible,
		}): INote => ({
			id,
			createdTimestamp: created_at,
			updatedTimestamp: updated_at,
			isSnapshotsDisabled: Boolean(history_disabled),
			isVisible: Boolean(visible),
			content: { title, text },
		}),
	);

function formatNoteMeta(meta: Partial<NoteMeta>) {
	return Object.fromEntries(
		Object.entries(meta).map((item): [string, string | number] => {
			const [key, value] = item as unknown as {
				[K in keyof NoteMeta]: [K, NoteMeta[K]];
			}[keyof NoteMeta];

			switch (key) {
				case 'isSnapshotsDisabled':
					return ['history_disabled', Number(value)];
				case 'isVisible':
					return ['visible', Number(value)];
			}
		}),
	);
}

function getFetchQuery(
	{
		select,
		workspace,
	}: {
		select: Query;
		workspace?: string;
	},
	{ limit, page, tags = [], meta, sort }: NotesControllerFetchOptions = {},
) {
	if (page !== undefined && page < 1)
		throw new TypeError('Page value must not be less than 1');

	const metaEntries = Object.entries(
		formatNoteMeta({
			isVisible: true,
			...meta,
		}),
	);

	const sortFieldMap: { [K in NoteSortField]: string } = {
		id: 'id',
		createdAt: 'created_at',
		updatedAt: 'updated_at',
	};

	return qb.line(
		qb.sql`SELECT ${select} FROM notes`,
		qb
			.where(workspace ? qb.sql`workspace_id=${workspace}` : undefined)
			.and(
				tags.length === 0
					? undefined
					: qb.sql`id IN (SELECT target FROM attached_tags WHERE source IN (${qb.values(
							tags,
					  )}))`,
			)
			.and(...metaEntries.map(([key, value]) => qb.sql`${qb.raw(key)} = ${value}`)),
		sort
			? qb.line(
					qb.sql`ORDER BY`,
					qb.set([
						sort
							? qb.line(
									sortFieldMap[sort.by],
									sort.order === 'desc' ? 'DESC' : 'ASC',
							  )
							: undefined,
					]),
			  )
			: undefined,
		limit ? qb.limit(limit) : undefined,
		page && limit ? qb.offset((page - 1) * limit) : undefined,
	);
}

/**
 * Synced notes registry
 */
export class NotesController implements INotesController {
	private db;
	private readonly workspace;
	constructor(db: PGLiteDatabase, workspace: string) {
		this.db = db;
		this.workspace = workspace;
	}

	public async getById(id: NoteId): Promise<INote | null> {
		const db = wrapDB(this.db.get());

		const {
			rows: [note],
		} = await db.query(
			qb.sql`SELECT * FROM notes WHERE workspace_id=${this.workspace} AND id=${id}`,
			RowScheme,
		);

		return note ?? null;
	}

	public async getLength(query: NotesControllerFetchOptions = {}): Promise<number> {
		const db = wrapDB(this.db.get());

		const {
			rows: [{ count }],
		} = await db.query(
			getFetchQuery(
				{ select: qb.sql`COUNT(*) as count`, workspace: this.workspace },
				query,
			),
			z.object({ count: z.number() }),
		);

		return count;
	}

	public async get(query: NotesControllerFetchOptions = {}): Promise<INote[]> {
		const db = wrapDB(this.db.get());

		const { rows } = await db.query(
			getFetchQuery({ select: qb.sql`*`, workspace: this.workspace }, query),
			RowScheme,
		);

		return rows;
	}

	public async add(note: INoteContent, meta?: Partial<NoteMeta>): Promise<NoteId> {
		const db = wrapDB(this.db.get());

		const creationTime = new Date().getTime();

		// Insert data
		const metaEntries = Object.entries(formatNoteMeta(meta ?? {}));

		const {
			rows: [id],
		} = await db.query(
			qb.sql`INSERT INTO notes (${qb.set([
				qb.sql`"workspace_id","title","text","created_at","updated_at"`,
				...(metaEntries ? metaEntries.map(([key]) => key) : []),
			])}) VALUES (${qb.values([
				this.workspace,
				note.title,
				note.text,
				creationTime,
				creationTime,
				...metaEntries.map(([_key, value]) => value),
			])}) RETURNING id`,
			z.object({ id: z.string() }).transform(({ id }) => id),
		);

		return id;
	}

	public async update(id: string, updatedNote: INoteContent) {
		const db = wrapDB(this.db.get());

		const updateTime = new Date().getTime();
		const result = await db.query(
			qb.line(
				'UPDATE notes SET',
				qb.values({
					title: updatedNote.title,
					text: updatedNote.text,
					// TODO: should we really update note by update meta?
					updated_at: updateTime,
				}),
				qb.sql`WHERE id=${id} AND workspace_id=${this.workspace}`,
			),
		);

		if (!result.affectedRows || result.affectedRows < 1) {
			throw new Error('Note did not updated');
		}
	}

	public async delete(ids: NoteId[]): Promise<void> {
		const db = wrapDB(this.db.get());

		const { affectedRows = 0 } = await db.query(
			qb.sql`DELETE FROM notes WHERE workspace_id=${
				this.workspace
			} AND id IN (${qb.values(ids)})`,
		);

		if (affectedRows !== ids.length) {
			console.warn(
				`Not match deleted entries length. Expected: ${ids.length}; Deleted: ${affectedRows}`,
			);
		}
	}

	public async updateMeta(ids: NoteId[], meta: Partial<NoteMeta>): Promise<void> {
		const db = wrapDB(this.db.get());

		const { affectedRows = 0 } = await db.query(
			qb.sql`UPDATE notes SET ${qb.values(formatNoteMeta(meta))}
				WHERE workspace_id=${this.workspace} AND id IN (${qb.values(ids)})`,
		);

		if (affectedRows !== ids.length) {
			console.warn(
				`Not match updated entries length. Expected: ${ids.length}; Updated: ${affectedRows}`,
			);
		}
	}
}
