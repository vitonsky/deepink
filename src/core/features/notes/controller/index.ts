import { INote, INoteContent, NoteId } from '..';

export type NoteMeta = {
	isSnapshotsDisabled: boolean;
	isVisible: boolean;
};

export type NoteSortField = 'id' | 'createdAt' | 'updatedAt';

export type NotesControllerFetchOptions = {
	/**
	 * Limit notes
	 *
	 * @default 100
	 */
	limit?: number;

	/**
	 * Page number start of 1
	 *
	 * Implementations must thrown error for values less than 1
	 *
	 * @default 1
	 */
	page?: number;

	/**
	 * Filter notes by tags
	 */
	tags?: string[];

	/**
	 * Filters by note meta info
	 */
	meta?: Partial<NoteMeta>;

	search?: {
		text: string;
	};

	/**
	 * Sorting options
	 */
	sort?: {
		by: NoteSortField;
		order?: 'desc' | 'asc';
	};
};

/**
 * Notes controller interface
 */
export interface INotesController {
	/**
	 * Get note by ID
	 */
	getById(id: NoteId): Promise<INote | null>;

	/**
	 * Get number of notes
	 */
	getLength(query?: NotesControllerFetchOptions): Promise<number>;

	/**
	 * Primary method to get notes filtered by parameters
	 */
	get(options?: NotesControllerFetchOptions): Promise<INote[]>;

	/**
	 * Create note and return unique id of new note
	 */
	add(note: INoteContent, meta?: Partial<NoteMeta>): Promise<NoteId>;

	/**
	 * Update note by unique id
	 */
	update(id: NoteId, updatedNote: INoteContent): Promise<void>;

	/**
	 * Update note meta information
	 */
	updateMeta(ids: NoteId[], meta: Partial<NoteMeta>): Promise<void>;

	/**
	 * Deletes all notes with specified ids
	 */
	delete(ids: NoteId[]): Promise<void>;

	/**
	 * Updates the status of specific notes
	 */
	updateStatus(ids: NoteId[], status: { deleted: boolean }): Promise<void>;
}
