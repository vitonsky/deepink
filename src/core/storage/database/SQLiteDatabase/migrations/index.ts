import { Database } from 'better-sqlite3';

export type MigrationsTarget = Database;

const migrations = [
	{
		version: 1,
		up: async (db: MigrationsTarget) => {
			db.exec(`CREATE TABLE "files" (
				"id"	TEXT NOT NULL UNIQUE,
				"name"	TEXT NOT NULL,
				"mimetype"	TEXT NOT NULL,
				PRIMARY KEY("id")
			)`);
		},
	},
	{
		version: 2,
		up: async (db: MigrationsTarget) => {
			db.exec(`CREATE TABLE "attachments" (
				"id"	TEXT NOT NULL UNIQUE,
				"file"	TEXT NOT NULL,
				"note"	TEXT NOT NULL,
				PRIMARY KEY("id")
			)`);
		},
	},
	{
		version: 3,
		up: async (db: MigrationsTarget) => {
			db.transaction(() => {
				db.exec('DROP TABLE "attachments"');
				db.exec(`CREATE TABLE "attachments" (
					"id"	TEXT NOT NULL UNIQUE,
					"file"	TEXT NOT NULL,
					"note"	TEXT NOT NULL,
					PRIMARY KEY("id")
					UNIQUE(file,note)
					)`);
			})();
		},
	},
	{
		version: 4,
		up: async (db: MigrationsTarget) => {
			db.transaction(() => {
				db.exec(`CREATE TABLE "workspaces" (
					"id"	TEXT NOT NULL UNIQUE,
					"name"	TEXT NOT NULL,
					PRIMARY KEY("id")
					)`);
			})();
		},
	},
	{
		version: 5,
		up: async (db: MigrationsTarget) => {
			db.transaction(() => {
				db.exec(`CREATE TABLE "note_versions" (
					"id" TEXT NOT NULL UNIQUE,
					"note_id" TEXT NOT NULL,
					"created_at" INTEGER NOT NULL DEFAULT 0,
					"title" TEXT NOT NULL,
					"text" TEXT NOT NULL,
					PRIMARY KEY("id"),
					FOREIGN KEY (note_id) REFERENCES notes(id) ON DELETE CASCADE
				)`);
			})();
		},
	},
	{
		version: 6,
		up: async (db: MigrationsTarget) => {
			db.transaction(() => {
				db.exec(
					`ALTER TABLE "notes" ADD COLUMN "isSnapshotsDisabled" INTEGER NOT NULL DEFAULT 0;`,
				);
			})();
		},
	},
	{
		version: 7,
		up: async (db: MigrationsTarget) => {
			db.transaction(() => {
				db.exec(
					`ALTER TABLE "notes" ADD COLUMN "isVisible" INTEGER NOT NULL DEFAULT 1;`,
				);
			})();
		},
	},
] as const;

export const latestSchemaVersion = migrations[migrations.length - 1].version;

export const migrateToLatestSchema = async (db: MigrationsTarget, verbose = false) => {
	let currentVersion = Number(db.pragma('main.user_version', { simple: true }));

	if (verbose) {
		console.log('DB version', currentVersion);
	}

	// TODO: ensure sorted versions
	for (const migration of migrations) {
		if (currentVersion >= migration.version) continue;

		const nextVersion = migration.version;

		await migration.up(db);
		db.pragma(`main.user_version = ${nextVersion};`);

		currentVersion = nextVersion;

		if (verbose) {
			console.log('DB version updated to', currentVersion);
		}
	}
};
