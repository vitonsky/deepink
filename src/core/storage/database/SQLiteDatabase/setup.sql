-- TODO: remove default value for `workspace_id`
CREATE TABLE "notes" (
	"id" TEXT NOT NULL UNIQUE,
	"workspace_id" TEXT NOT NULL DEFAULT 'none',
	"title" TEXT NOT NULL,
	"text" TEXT NOT NULL,
	"creationTime" INTEGER NOT NULL DEFAULT 0,
	"lastUpdateTime" INTEGER NOT NULL DEFAULT 0,
	"isSnapshotsDisabled" INTEGER NOT NULL DEFAULT 0,
	PRIMARY KEY("id")
);
-- Note text snapshots
CREATE TABLE "note_versions" (
	"id" TEXT NOT NULL UNIQUE,
	"note_id" TEXT NOT NULL,
	"created_at" INTEGER NOT NULL DEFAULT 0,
	"title" TEXT NOT NULL,
	"text" TEXT NOT NULL,
	PRIMARY KEY("id"),
	FOREIGN KEY (note_id) REFERENCES notes(id) ON DELETE CASCADE
);
CREATE TABLE "files" (
	"id" TEXT NOT NULL UNIQUE,
	"workspace_id" TEXT NOT NULL DEFAULT 'none',
	"name" TEXT NOT NULL,
	"mimetype" TEXT NOT NULL,
	PRIMARY KEY("id")
);
CREATE TABLE "attachments" (
	"id" TEXT NOT NULL UNIQUE,
	"workspace_id" TEXT NOT NULL DEFAULT 'none',
	"file" TEXT NOT NULL,
	"note" TEXT NOT NULL,
	PRIMARY KEY("id") UNIQUE(file, note)
);
CREATE TABLE "tags" (
	"id" TEXT NOT NULL UNIQUE,
	"workspace_id" TEXT NOT NULL DEFAULT 'none',
	"name" TEXT NOT NULL,
	"parent" TEXT,
	PRIMARY KEY("id") UNIQUE(name, parent)
);
CREATE TABLE "attachedTags" (
	"id" TEXT NOT NULL UNIQUE,
	"workspace_id" TEXT NOT NULL DEFAULT 'none',
	"source" TEXT NOT NULL,
	"target" TEXT NOT NULL,
	PRIMARY KEY("id") UNIQUE(source, target)
);
CREATE TABLE "workspaces" (
	"id" TEXT NOT NULL UNIQUE,
	"name" TEXT NOT NULL,
	PRIMARY KEY("id")
);