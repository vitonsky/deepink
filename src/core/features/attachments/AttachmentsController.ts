import { v4 as uuid4 } from 'uuid';
import { z } from 'zod';
import { qb } from '@utils/db/query-builder';
import { wrapDB } from '@utils/db/wrapDB';

import { SQLiteDatabase } from '../../storage/database/SQLiteDatabase/SQLiteDatabase';

/**
 * attachments manager, to track attachments usage and to keep consistency
 */
export class AttachmentsController {
	private db;
	private readonly workspace;
	constructor(db: SQLiteDatabase, workspace: string) {
		this.db = db;
		this.workspace = workspace;
	}

	public async set(targetId: string, attachments: string[]) {
		const db = this.db.get();

		if (attachments.length === 0) {
			db.prepare('DELETE FROM attachments WHERE workspace_id=? AND note=?').run(
				this.workspace,
				targetId,
			);
		} else {
			const valuesPlaceholders = attachments.map(() => `(?, ?, ?, ?)`).join(', ');

			const placeholdersData: string[] = [];
			attachments.forEach((attachmentId) => {
				placeholdersData.push(uuid4());
				placeholdersData.push(this.workspace);
				placeholdersData.push(targetId);
				placeholdersData.push(attachmentId);
			});

			db.transaction(() => {
				db.prepare('DELETE FROM attachments WHERE workspace_id=? AND note=?').run(
					this.workspace,
					targetId,
				);
				db.prepare(
					`INSERT INTO attachments ("id", "workspace_id", "note", "file") VALUES ${valuesPlaceholders};`,
				).run(placeholdersData);
			})();
		}
	}

	public async get(targetId: string): Promise<string[]> {
		const db = this.db.get();

		return db
			.prepare(
				'SELECT file FROM attachments WHERE workspace_id=? AND note=? ORDER BY rowid',
			)
			.all(this.workspace, targetId)
			.map((row) => (row as { file: string }).file);
	}

	public async delete(resources: string[]) {
		const db = this.db.get();

		const placeholders = Array(resources.length).fill('?').join(',');
		db.prepare(
			`DELETE FROM attachments WHERE workspace_id=? AND file IN (${placeholders})`,
		).run(this.workspace, ...resources);
	}

	public async query() {
		const db = wrapDB(this.db.get());

		return z
			.object({
				id: z.string(),
				file: z.string(),
				note: z.string(),
			})
			.array()
			.parse(
				db.all(
					qb.sql`SELECT id, file, note FROM attachments WHERE workspace_id=${this.workspace}`,
				),
			);
	}
}
