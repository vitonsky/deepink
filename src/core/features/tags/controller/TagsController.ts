import { createEvent } from 'effector';
import { v4 as uuid4 } from 'uuid';

import { SQLiteDatabase } from '../../../storage/database/SQLiteDatabase/SQLiteDatabase';

import tagsQuery from './selectTagsWithResolvedNames.sql';
import { IResolvedTag, ITag } from '..';

type ChangeEvent = 'tags' | 'noteTags';

export class TagsController {
	private readonly db;
	private readonly workspace;
	private readonly onChanged;
	constructor(db: SQLiteDatabase, workspace: string) {
		this.db = db;
		this.workspace = workspace;
		this.onChanged = createEvent<ChangeEvent>();
	}

	public onChange(event: (scope: ChangeEvent) => void) {
		const subscription = this.onChanged.watch(event);
		return () => subscription.unsubscribe();
	}

	/**
	 * Returns tags  list
	 */
	public async getTags(): Promise<IResolvedTag[]> {
		const db = this.db.get();

		return (
			db
				.prepare(`${tagsQuery} WHERE workspace_id=? ORDER BY rowid`)
				.all(this.workspace) as any[]
		).map(({ id, name, resolvedName, parent }) => ({
			id,
			name,
			resolvedName,
			parent: parent ?? null,
		}));
	}

	public async add(name: string, parent: null | string): Promise<string> {
		const db = this.db.get();

		let lastId: number;

		const segments = name.split('/');
		if (segments.length > 1) {
			const results = db.transaction(() =>
				segments.map((name, idx) => {
					const isFirstItem = idx === 0;

					if (isFirstItem) {
						return db
							.prepare(
								'INSERT INTO tags ("id", "workspace_id", "name", "parent") VALUES (?, ?, ?, ?)',
							)
							.run(uuid4(), this.workspace, name, parent);
					}

					return db
						.prepare(
							'INSERT INTO tags ("id", "workspace_id", "name", "parent") VALUES (?, ?, ?, (SELECT id FROM tags WHERE ROWID=last_insert_rowid()))',
						)
						.run(uuid4(), this.workspace, name);
				}),
			)();

			lastId = Number(results[results.length - 1].lastInsertRowid);
		} else {
			const insertResult = db
				.prepare(
					'INSERT INTO tags ("id", "workspace_id", "name", "parent") VALUES (?, ?, ?, ?)',
				)
				.run(uuid4(), this.workspace, name, parent);
			if (insertResult.lastInsertRowid === undefined) {
				throw new Error('Last insert id not found');
			}

			lastId = Number(insertResult.lastInsertRowid);
		}

		// Get generated id
		const selectWithId = (await db
			.prepare('SELECT `id` FROM tags WHERE rowid=?')
			.get(lastId)) as any;
		if (!selectWithId || !selectWithId.id) {
			throw new Error("Can't get id of inserted row");
		}

		this.onChanged('tags');

		return selectWithId.id;
	}

	public async update(tag: ITag): Promise<void> {
		const db = this.db.get();

		db.prepare('UPDATE tags SET name=?, parent=? WHERE id=? AND workspace_id=?').run(
			tag.name,
			tag.parent,
			tag.id,
			this.workspace,
		);

		this.onChanged('tags');
	}

	public async delete(id: string): Promise<void> {
		const db = this.db.get();

		const tagsIdForRemove = (
			db
				.prepare(
					`
			WITH RECURSIVE tagTree AS (
				SELECT
					id, parent, name, id AS root
				FROM tags
				UNION ALL
				SELECT
					t.id, t.parent, t.name, t2.root
				FROM tags t
				INNER JOIN tagTree t2
				ON t.parent = t2.id
			)
			SELECT id FROM tagTree WHERE root IN (?) GROUP BY id
		`,
				)
				.all(id) as { id: string }[]
		).map(({ id }) => id);

		db.transaction(() => {
			db.prepare(
				`DELETE FROM tags WHERE workspace_id=? AND id IN (${Array(
					tagsIdForRemove.length,
				)
					.fill('?')
					.join(',')})`,
			).run(this.workspace, ...tagsIdForRemove);

			db.prepare(
				`DELETE FROM attachedTags WHERE workspace_id=? AND source IN (${Array(
					tagsIdForRemove.length,
				)
					.fill('?')
					.join(',')})`,
			).run(this.workspace, ...tagsIdForRemove);
		})();

		this.onChanged('tags');
	}

	/**
	 * Returns tags attached to a entity
	 */
	public async getAttachedTags(target: string): Promise<IResolvedTag[]> {
		const db = this.db.get();

		const query = [
			tagsQuery,
			`WHERE t.id IN (SELECT source FROM attachedTags WHERE workspace_id=? AND target = ?)`,
		].join(' ');

		return (db.prepare(query).all(this.workspace, target) as any[]).map(
			({ id, name, resolvedName, parent }) => ({
				id,
				name,
				resolvedName,
				parent: parent ?? null,
			}),
		);
	}

	public async setAttachedTags(target: string, tags: string[]): Promise<void> {
		const db = this.db.get();

		db.transaction(() => {
			db.prepare(`DELETE FROM attachedTags WHERE workspace_id=? AND target=?`).run(
				this.workspace,
				target,
			);
			if (tags.length > 0) {
				db.prepare(
					`INSERT INTO attachedTags(id,workspace_id,source,target) VALUES ${Array(
						tags.length,
					)
						.fill('(?, ?, ?, ?)')
						.join(',')}`,
				).run(
					tags.map((tagId) => [uuid4(), this.workspace, tagId, target]).flat(),
				);
			}
		})();

		this.onChanged('noteTags');
	}
}
