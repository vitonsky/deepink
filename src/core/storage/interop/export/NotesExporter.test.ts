// @vitest-environment jsdom

import { AttachmentsController } from '@core/features/attachments/AttachmentsController';
import { createFileManagerMock } from '@core/features/files/__tests__/mocks/createFileManagerMock';
import { createTextFile } from '@core/features/files/__tests__/mocks/createTextFile';
import { FilesController } from '@core/features/files/FilesController';
import { formatNoteLink, formatResourceLink } from '@core/features/links';
import { NotesController } from '@core/features/notes/controller/NotesController';
import { TagsController } from '@core/features/tags/controller/TagsController';
import { createFileControllerMock } from '@utils/mocks/fileControllerMock';

import { openDatabase } from '../../database/SQLiteDatabase/SQLiteDatabase';

import { NotesExporter } from '.';

const FAKE_WORKSPACE_NAME = 'fake-workspace-id';

const dbFile = createFileControllerMock();
const dbPromise = openDatabase(dbFile);

afterAll(async () => {
	const db = await dbPromise;
	await db.close();
});

describe('Notes and files may be exported', () => {
	const fileManager = createFileManagerMock();

	test('Create notes', async () => {
		const db = await dbPromise;
		const notesRegistry = new NotesController(db, FAKE_WORKSPACE_NAME);
		const tagsRegistry = new TagsController(db, FAKE_WORKSPACE_NAME);

		const attachmentsRegistry = new AttachmentsController(db, FAKE_WORKSPACE_NAME);
		const filesRegistry = new FilesController(
			db,
			fileManager,
			attachmentsRegistry,
			FAKE_WORKSPACE_NAME,
		);

		// Upload files
		const file1Id = await filesRegistry.add(
			createTextFile('Text content of attachment'),
		);
		await Promise.all(
			Array(10)
				.fill(null)
				.map(() => filesRegistry.add(createTextFile('UNUSED FILE'))),
		);

		// Create tags
		const tags = await Promise.all([
			tagsRegistry.add('foo', null),
			tagsRegistry.add('bar', null),
			tagsRegistry.add('baz', null),
		]);

		// Add notes
		const note1Id = await notesRegistry.add({
			title: 'Note 1',
			text: 'Hello world!',
		});

		await tagsRegistry.setAttachedTags(note1Id, tags);

		await notesRegistry.add({
			title: 'Note 2',
			text: `Mention for [Note 1](${formatNoteLink(note1Id)})`,
		});

		await attachmentsRegistry.set(
			await notesRegistry.add({
				title: 'Note 3',
				text: `Note with [attachment](${formatResourceLink(file1Id)})`,
			}),
			[file1Id],
		);

		await attachmentsRegistry.set(
			await notesRegistry.add({
				title: 'Note 4',
				text: `Note with [attachment](${formatResourceLink(
					file1Id,
				)}) and mention for [Note 1](${formatNoteLink(note1Id)})`,
			}),
			[file1Id],
		);
		await attachmentsRegistry.set(
			await notesRegistry.add({
				title: 'Note 5',
				text: `Note with [invalid attachment](${formatResourceLink(
					'00000000-0000-0000-0000-000000000000',
				)}) and mention for [invalid note](${formatNoteLink(
					'00000000-0000-0000-0000-000000000000',
				)})`,
			}),
			[file1Id],
		);
	});

	test('Export notes', async () => {
		const db = await dbPromise;
		const notesRegistry = new NotesController(db, FAKE_WORKSPACE_NAME);
		const tagsRegistry = new TagsController(db, FAKE_WORKSPACE_NAME);

		const attachmentsRegistry = new AttachmentsController(db, FAKE_WORKSPACE_NAME);
		const filesRegistry = new FilesController(
			db,
			fileManager,
			attachmentsRegistry,
			FAKE_WORKSPACE_NAME,
		);

		const saveFile = vi.fn();

		const exporter = new NotesExporter({
			saveFile,
			filesRegistry,
			notesRegistry,
			tagsRegistry,
		});

		// TODO: Replace resource URLs like `Note with [attachment](res://[REDACTED-UUID])` to file paths
		// TODO: save notes one by one
		// Exported notes
		await expect(
			exporter.exportNotes().then((notes) =>
				notes.map((note, index) => {
					note.id = '[REDACTED UUID]';
					note.data = note.data
						.replace(/created: \d+/, `created: ${100_000 + index}`)
						.replace(/updated: \d+/, `updated: ${200_000 + index}`)
						.replaceAll(/[a-z\d]+(-[a-z\d]+){4}/g, '[REDACTED-UUID]');
					return note;
				}),
			),
		).resolves.toMatchSnapshot('Exported notes array');

		// TODO: callback must be called only once per unique file (not to save twice the same file)
		// Callback calls to save files
		expect(saveFile.mock.calls).toEqual([
			[expect.any(File), expect.any(String)],
			[expect.any(File), expect.any(String)],
		]);
	});
});
