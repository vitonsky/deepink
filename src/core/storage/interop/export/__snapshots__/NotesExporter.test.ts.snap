// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Notes and files may be exported > Export notes > Exported notes array 1`] = `
[
  {
    "data": "---
title: Note 1
created: 100000
updated: 200000
tags:
  - bar
  - baz
  - foo

---

Hello world!
",
    "id": "[REDACTED UUID]",
  },
  {
    "data": "---
title: Note 2
created: 100001
updated: 200001
tags: []

---

Mention for [Note 1](./[REDACTED-UUID].md)
",
    "id": "[REDACTED UUID]",
  },
  {
    "data": "---
title: Note 3
created: 100002
updated: 200002
tags: []

---

Note with [attachment](res://[REDACTED-UUID])
",
    "id": "[REDACTED UUID]",
  },
  {
    "data": "---
title: Note 4
created: 100003
updated: 200003
tags: []

---

Note with [attachment](res://[REDACTED-UUID]) and mention for [Note 1](./[REDACTED-UUID].md)
",
    "id": "[REDACTED UUID]",
  },
  {
    "data": "---
title: Note 5
created: 100004
updated: 200004
tags: []

---

Note with [invalid attachment](res://[REDACTED-UUID]) and mention for [invalid note](./[REDACTED-UUID].md)
",
    "id": "[REDACTED UUID]",
  },
]
`;
